### PR TITLE
[WIP] Import ast from json format

### PR DESCRIPTION
**Pull request by djudjuu**

### CODE REVIEW

File: libsolidity/ast/ASTJsonConverter.cpp [hunk 1]
```diff
@@ -252,6 +252,7 @@ bool ASTJsonConverter::visit(ContractDefinition const& _node)
 {
 	setJsonNode(_node, "ContractDefinition", {
 		make_pair("name", _node.name()),
+		make_pair("documentation", _node.documentation() ? Json::Value(*_node.documentation()) : Json::nullValue),
```
Comment (axic) [inline libsolidity/ast/ASTJsonConverter.cpp]: Can you pull out the documentation part into a separate PR? There is an issue for that and would be nice to close it off separately.

File: libsolidity/ast/ASTJsonConverter.cpp [hunk 2]
```diff
@@ -704,13 +705,13 @@ bool ASTJsonConverter::visit(ElementaryTypeNameExpression const& _node)
...
 {
-	char const* tokenString = Token::toString(_node.token());
+	string tokenString = tokenKind(_node.token());
 	Json::Value value{_node.value()};
 	if (!dev::validateUTF8(_node.value()))
 		value = Json::nullValue;
 	Token::Value subdenomination = Token::Value(_node.subDenomination());
 	std::vector<pair<string, Json::Value>> attributes = {
-		make_pair("token", tokenString ? tokenString : Json::Value()),
+		make_pair("kind", tokenString),
```
Comment (axic) [inline libsolidity/ast/ASTJsonConverter.cpp]: This change seems to be a bugfix to the current printing. Perhaps we could pull it into a separate PR and merge it now?

File: libsolidity/interface/CompilerStack.cpp [hunk 3]
```diff
@@ -807,25 +826,26 @@ string CompilerStack::createOnChainMetadata(Contract const& _contract) const
...
 	meta["sources"] = Json::objectValue;
-	for (auto const& s: m_sources)
-	{
-		solAssert(s.second.scanner, "Scanner not available");
-		meta["sources"][s.first]["keccak256"] =
-			"0x" + toHex(dev::keccak256(s.second.scanner->source()).asBytes());
-		if (m_metadataLiteralSources)
-			meta["sources"][s.first]["content"] = s.second.scanner->source();
-		else
+	if (!m_importedSources)
```
Comment (axic) [inline libsolidity/interface/CompilerStack.cpp]: The AST JSON should be placed included here. (See #2166.)

File: libsolidity/ast/ASTJsonConverter.cpp (hunk 1 above)
Comment (djudjuu) [reply libsolidity/ast/ASTJsonConverter.cpp]: done.

File: libsolidity/ast/ASTJsonConverter.cpp (hunk 2 above)
Comment (djudjuu) [reply libsolidity/ast/ASTJsonConverter.cpp]: yeah...planned to do that anyway, but i needed it to progress...

File: libsolidity/ast/ASTJsonImporter.h [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+#include <libsolidity/ast/ASTAnnotations.h>
+#include <libevmasm/SourceLocation.h>
+
+namespace dev
+{
+namespace solidity
+{
+
+/**
+ * takes an AST in Json Format and recreates it with AST-Nodes
```
Comment (chriseth) [inline libsolidity/ast/ASTJsonImporter.h]: Please start full sentences with a capital letter and end them with a full-stop :-)

File: libsolidity/ast/ASTJsonImporter.h [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+ */
+class ASTJsonImporter
+{
+public:
+	/// Create an Importer to import a given abstract syntax tree in Json format to an ASTNode
+	ASTJsonImporter(std::map<std::string, Json::Value const*> _sourceList);
+
+        ///this is the primary Interface from the Outside
```
Comment (chriseth) [inline libsolidity/ast/ASTJsonImporter.h]: Would it be better to have just a single static function and a private constructor? Is there a benefit of having a two-step process?

File: libsolidity/ast/ASTJsonImporter.h [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+	ASTPointer<NewExpression> createNewExpression(Json::Value const& _node);
+	ASTPointer<MemberAccess> createMemberAccess(Json::Value const& _node);
+	ASTPointer<IndexAccess> createIndexAccess(Json::Value const& _node);
+	ASTPointer<Identifier> createIdentifier(SourceLocation location, std::string const& name);
+	ASTPointer<ElementaryTypeNameExpression> createElementaryTypeNameExpression(Json::Value const& _node);
+	ASTPointer<ASTNode> createLiteral(Json::Value const& _node);
+
+	template<class T>
+	ASTPointer<T> nullOrCast(Json::Value _json);
+	Declaration::Visibility getVisibility(Json::Value const& _node);
```
Comment (chriseth) [inline libsolidity/ast/ASTJsonImporter.h]: `get` as a prefix is discouraged (I know it is present in lots of places, but I think it just adds verbosity).

File: libsolidity/ast/ASTJsonImporter.cpp [hunk 7]
```diff
@@ -0,0 +1,1003 @@
...
+		m_currentSource = srcPair.first;
+		m_sourceUnits[srcPair.first] =  createSourceUnit(*srcPair.second);
+	}
+	return m_sourceUnits;
+}
+
+SourceLocation const ASTJsonImporter::createSourceLocation(Json::Value const& _node)
+{
+	solAssert(!_node["src"].isNull(), "JsonValue should not be an ASTNode");
+        string srcString = _node["src"].asCString();
```
Comment (chriseth) [inline libsolidity/ast/ASTJsonImporter.cpp]: Please use tabs for indentation.

File: libsolidity/ast/ASTJsonImporter.cpp [hunk 8]
```diff
@@ -0,0 +1,1003 @@
...
+{
+	dev::solidity::ASTPointer<T> ret = dynamic_pointer_cast<T>(_ast);
+	if (!ret)
+		solAssert(false, "");
+	else
+		return ret;
+}
+}
+
+namespace dev
```
Comment (chriseth) [inline libsolidity/ast/ASTJsonImporter.cpp]: The style is to put declarations inside namespaces, but not the definitions, i.e. use `namespace { ...}` in header files and `using namespace dev; using namespace dev::solidity;` in cpp files.

File: libsolidity/ast/ASTJsonImporter.cpp (hunk 7 above)
Comment (chriseth) [inline libsolidity/ast/ASTJsonImporter.cpp]: please use `asString()` instead of `asCString()`

File: libsolidity/ast/ASTJsonImporter.cpp [hunk 9]
```diff
@@ -0,0 +1,1003 @@
...
+	return m_sourceUnits;
+}
+
+SourceLocation const ASTJsonImporter::createSourceLocation(Json::Value const& _node)
+{
+	solAssert(!_node["src"].isNull(), "JsonValue should not be an ASTNode");
+        string srcString = _node["src"].asCString();
+        vector<string> pos;
+        boost::algorithm::split(pos, srcString, boost::is_any_of(":"));
+        int start = stoi(pos[0]);
```
Comment (chriseth) [inline libsolidity/ast/ASTJsonImporter.cpp]: This throws an exception, please catch it at the appropriate point (might be as high up as the entry function into this class).

File: libsolidity/ast/ASTJsonImporter.cpp [hunk 10]
```diff
@@ -0,0 +1,1003 @@
...
+	int end = start + stoi(pos[1]);
```
Comment (chriseth) [inline libsolidity/ast/ASTJsonImporter.cpp]: please `solAssert` that there are enough elements in `pos`

File: libsolidity/ast/ASTJsonImporter.cpp [hunk 11]
```diff
@@ -0,0 +1,1003 @@
...
+	shared_ptr<string const> source = make_shared<string const>(m_currentSource);
```
Comment (chriseth) [inline libsolidity/ast/ASTJsonImporter.cpp]: It does not have to be the current source, but rather it is an index into the source index list.

File: libsolidity/ast/ASTJsonImporter.cpp [hunk 12]
```diff
@@ -0,0 +1,1003 @@
...
+                end,
+                source
+        );
+}
+
+//helper functions
+template<class T>
+ASTPointer<T> ASTJsonImporter::nullOrCast(Json::Value _json)
+{
+	if(!_json.isNull())
```
Comment (chriseth) [inline libsolidity/ast/ASTJsonImporter.cpp]: Please swap the two cases to avoid the `!`

File: libsolidity/ast/ASTJsonImporter.cpp [hunk 13]
```diff
@@ -0,0 +1,1003 @@
...
+		return castPointer<T>(convertJsonToASTNode(_json));
+	else
+		return nullptr;
+}
+
+Token::Value ASTJsonImporter::scanSingleToken(Json::Value _node)
+{
+        Token::Value ret;
```
Comment (chriseth) [inline libsolidity/ast/ASTJsonImporter.cpp]: Please don't use spaces.

File: libsolidity/ast/ASTJsonImporter.cpp [hunk 14]
```diff
@@ -0,0 +1,1003 @@
...
+	Scanner scanner(CharStream(_node.asCString()));
+	ret = scanner.currentToken();
+	if (scanner.next() == Token::EOS)
+		return std::move(ret);
```
Comment (chriseth) [inline libsolidity/ast/ASTJsonImporter.cpp]: `std::move` should not be used with `return`

File: libsolidity/ast/ASTJsonImporter.cpp [hunk 15]
```diff
@@ -0,0 +1,1003 @@
...
+		return castPointer<T>(convertJsonToASTNode(_json));
+	else
+		return nullptr;
+}
+
+Token::Value ASTJsonImporter::scanSingleToken(Json::Value _node)
+{
+        Token::Value ret;
+	Scanner scanner(CharStream(_node.asCString()));
+	ret = scanner.currentToken();
```
Comment (chriseth) [inline libsolidity/ast/ASTJsonImporter.cpp]: Does it make sense to have the `ret` variable at all?

File: libsolidity/ast/ASTJsonImporter.cpp [hunk 16]
```diff
@@ -0,0 +1,1003 @@
...
+		return std::move(ret);
+        else
+	    BOOST_THROW_EXCEPTION(InvalidAstError() << errinfo_comment("Invalid AST entry."));
+}
+
+ASTPointer<ASTNode> ASTJsonImporter::convertJsonToASTNode(Json::Value const& _json)
+{
+	if (_json["nodeType"] == "SourceUnit")
+	    return createSourceUnit(_json);
+	if (_json["nodeType"].asString() == "PragmaDirective")
```
Comment (chriseth) [inline libsolidity/ast/ASTJsonImporter.cpp]: Why do some of them have `asString` and some not? Also it might be better to store `_json["nodeType"].asString()` in a local variable first.

File: libsolidity/ast/ASTJsonImporter.cpp [hunk 17]
```diff
@@ -0,0 +1,1003 @@
...
+
+ASTPointer<SourceUnit> ASTJsonImporter::createSourceUnit(Json::Value const& _node)
+{
+	SourceLocation const& location = createSourceLocation(_node);
+	vector<ASTPointer<ASTNode>> nodes;
+	for (auto& child: _node["nodes"])
+		nodes.push_back(convertJsonToASTNode(child));
+	ASTPointer<SourceUnit> tmp = make_shared<SourceUnit>(location, nodes);
+	tmp->annotation().path = m_currentSource;
+	tmp->setID(_node["id"].asInt());
```
Comment (chriseth) [inline libsolidity/ast/ASTJsonImporter.cpp]: Since `id` and `location` is part of every ast node, it might make sense to create a template function:
```
template <typename T, typename... Args>
ASTPointer<T> createASTNode(Json::Value const& _node, Args&&... _args)
{
    auto n = make_shared<T>(createSourceLocation(_node), forward<Args>(_args)...);
    n->setID(_node["id"].asInt());
    return n;
}
```

and then you would use it as:

```
ASTPointer<SourceUnit> ASTJsonImporter::createSourceUnit(Json::Value const& _node)
{
	vector<ASTPointer<ASTNode>> nodes;
	for (auto& child: _node["nodes"])
		nodes.emplace_back(convertJsonToASTNode(child));
	ASTPointer<SourceUnit> tmp = createASTNode<SourceUnit>(_node, nodes);
	tmp->annotation().path = m_currentSource;
        return tmp;
}
```

File: libsolidity/ast/ASTJsonImporter.cpp [hunk 18]
```diff
@@ -0,0 +1,1003 @@
...
+{
+	vector<Token::Value> tokens;
+	vector<ASTString> literals;
+	for (auto const& lit: _node["literals"])
+	{
+                string l = lit.asCString();
+                literals.push_back(l);
+                tokens.push_back(scanSingleToken(l));
+	}
+	SourceLocation const& location = createSourceLocation(_node);
```
Comment (chriseth) [inline libsolidity/ast/ASTJsonImporter.cpp]: You cannot use `const&` here.

File: solc/CommandLineInterface.cpp [hunk 19]
```diff
@@ -760,8 +763,78 @@ bool CommandLineInterface::processInput()
...
+						supplementarySourceCodes[srcPair.first] = srcPair.second;
+				}
+			}
+			//replace the internal map so that every source has its own entry
+			m_sourceCodes = tmp_sources;
+			//feed AST to compiler
+			m_compiler->reset(false);
+			bool import = m_compiler->importASTs(sourceJsons);
+			if (supplementarySourceCodes.size() > 0)
+				m_compiler->saveImportedSourceCodes(supplementarySourceCodes);
```
Comment (chriseth) [inline solc/CommandLineInterface.cpp]: Do we really have to distinguish between the source files whensupplied to the "create from json ast" mode and the regular source files?

File: solc/CommandLineInterface.cpp [hunk 20]
```diff
@@ -760,8 +763,78 @@ bool CommandLineInterface::processInput()
...
+			{
+				Json::Value* ast = new Json::Value(); //use shared-Pointer here?
+				// try parsing as json
+				if (reader.parse(srcPair.second, *ast, false))
+				{
+//					BOOST_THROW_EXCEPTION(InvalidAstError() << errinfo_comment("could not be parsed to JSON")); //TODO what to throw
+					// case a)
+					if ((*ast).isMember("sourceList") && (*ast).isMember("sources"))
+					{
+					for (auto& src: (*ast)["sourceList"]) //check here for ast-correctness as well
```
Comment (pirapira) [inline solc/CommandLineInterface.cpp]: The indentation looks off.

File: solc/CommandLineInterface.cpp [hunk 21]
```diff
@@ -760,8 +763,78 @@ bool CommandLineInterface::processInput()
...
+		{
+			//read file contents, which either are
+			// a) a json-file with multiple sources
+			// b) a json-file with one source only
+			// c) a .sol file that corresponds to a given json
+			Json::Reader reader;
+			map<string, Json::Value const*> sourceJsons; // will be given to the compilerimportfunction
+			map<string, string> tmp_sources; //used to generate the onchainmetadata-hash
+			map<string, string> supplementarySourceCodes;
+			for (auto& srcPair: m_sourceCodes) // aka <string, string>
```
Comment (pirapira) [inline solc/CommandLineInterface.cpp]: Have you tried `const` here?

File: solc/CommandLineInterface.cpp (hunk 19 above)
Comment (djudjuu) [reply solc/CommandLineInterface.cpp]: when we wanted to add the .sol file in the metadata, we would have had to...Now that we decided to use the printed JSON as sourcefile, not anymore.

File: libsolidity/ast/ASTJsonImporter.cpp (hunk 17 above)
Comment (djudjuu) [reply libsolidity/ast/ASTJsonImporter.cpp]: good point, thx

File: libsolidity/ast/ASTJsonImporter.cpp (hunk 15 above)
Comment (djudjuu) [reply libsolidity/ast/ASTJsonImporter.cpp]: i figured i need to store it somewhere in order to check the next token for being EOS...is there a way i can return it conditionally on the next token without storing it?

Comment (axic) [reply libsolidity/ast/ASTJsonImporter.cpp]: `peekToken` or `peekNextToken` should do it

File: libsolidity/ast/ASTJsonImporter.h (hunk 5 above)
Comment (djudjuu) [reply libsolidity/ast/ASTJsonImporter.h]: I guess there is no benefit. I just modelled it upon the ASTJsonConverter...

File: solc/CommandLineInterface.cpp [hunk 22]
```diff
@@ -760,8 +763,68 @@ bool CommandLineInterface::processInput()
...
+						for (auto& src: (*ast)["sourceList"])
+						{
+							solAssert(
+								strcmp(
+									(*ast)["sources"][src.asCString()]["AST"]["nodeType"].asCString(),
+									"SourceUnit"
+								) == 0,
+								"the top-level node of the AST needs to be a 'SourceUnit'"
+							);
+							sourceJsons[src.asCString()] = &((*ast)["sources"][src.asString()]["AST"]);
```
Comment (axic) [inline solc/CommandLineInterface.cpp]: Why is this using `asCString`, while the other side is `asString`?

File: libsolidity/interface/CompilerStack.cpp [hunk 23]
```diff
@@ -807,26 +829,37 @@ string CompilerStack::createOnChainMetadata(Contract const& _contract) const
...
+			{
+				meta["sources"][s.first]["urls"] = Json::arrayValue;
+				meta["sources"][s.first]["urls"].append(
+					"bzzr://" + toHex(dev::swarmHash(s.second.scanner->source()).asBytes())
+				);
+			}
+
+		}
 		else
+			meta[s.first]["keccak256"] = "no source code available";
```
Comment (axic) [inline libsolidity/interface/CompilerStack.cpp]: This shouldn't be set to a non-hash. It is an optional field so can be omitted.

File: libsolidity/interface/CompilerStack.cpp [hunk 24]
```diff
@@ -807,26 +829,37 @@ string CompilerStack::createOnChainMetadata(Contract const& _contract) const
...
+		if (m_importedSources)
 		{
-			meta["sources"][s.first]["urls"] = Json::arrayValue;
-			meta["sources"][s.first]["urls"].append(
-				"bzzr://" + toHex(dev::swarmHash(s.second.scanner->source()).asBytes())
-			);
+			solAssert(m_sourceJsons.count(s.first), "no JSON found for source");
```
Comment (axic) [inline libsolidity/interface/CompilerStack.cpp:845]: For AST it should set:
```
meta["sources"][""]["content"] = dev::jsonCompactPrint(<ast>);
```

File: libsolidity/interface/CompilerStack.cpp [hunk 25]
```diff
@@ -807,26 +829,37 @@ string CompilerStack::createOnChainMetadata(Contract const& _contract) const
...
+//			meta["sources"][s.first]["AST-JSON"] = *(m_sourceJsons[sx.first]); //TODO?
 		}
 	}
 	meta["settings"]["optimizer"]["enabled"] = m_optimize;
+	meta["settings"]["compiledFrom"] = m_importedSources ? "AST-Json" : "Solidity sourcefile";
```
Comment (axic) [inline libsolidity/interface/CompilerStack.cpp]: This is not needed given the `language` is set.

File: libsolidity/interface/CompilerStack.cpp (hunk 24 above)
Comment (axic) [reply libsolidity/interface/CompilerStack.cpp:845]: Sorry realised you do support multiple filenames for the AST, so the filename should be present.

File: libsolidity/ast/ASTJsonImporter.cpp [hunk 26]
```diff
@@ -0,0 +1,925 @@
...
+
+using namespace std;
+
+namespace
+{
+template<class T>
+dev::solidity::ASTPointer<T> castPointer(dev::solidity::ASTPointer<dev::solidity::ASTNode> _ast)
+{
+	dev::solidity::ASTPointer<T> ret = dynamic_pointer_cast<T>(_ast);
+	if (!ret)
```
Comment (axic) [inline libsolidity/ast/ASTJsonImporter.cpp]: Could be replaced with
```
solAssert(ret, "");
return ret;
```

File: solc/CommandLineInterface.cpp (hunk 22 above)
Comment (djudjuu) [reply solc/CommandLineInterface.cpp]: there is no meaningful answer to this....maybe because of the entropy of the universe is increasing

**Comment by djudjuu** [comment]
currently, the analysis step breaks when trying to resolveNamesAndTypes() (compilerStack.cpp : 204).
probably because the shared pointers aren't right i.e. the shared pointers referencing the ASTNodes (e.g. contractdefinition) have a count of 1 (entering the analyze() function after importASTS()), whereas the count is ~21 when entering analyze() after regular parsing.

note: compiling/importing a simple contract like
`contract D  {
  uint a = 5 wei;
  function foo(uint x) returns (uint)  {    a +=1 ;    return a;  }
}`

my guess right now is that my ASTJsonImporter::createSourceLocation() is at fault...
update: all good. found the bug. createIdentifier used a wrong name, so that references could not be resolved....

File: libsolidity/ast/ASTJsonImporter.cpp [hunk 27]
```diff
@@ -47,7 +47,6 @@ using namespace solidity;
 template <typename T, typename... Args>
 ASTPointer<T> ASTJsonImporter::createASTNode(Json::Value const& _node, Args&&... _args)
 {
-	astAssert(_node.isMember("id") && _node.isMember("src"), "'id' or 'src' field missing");
```
Comment (chriseth) [inline libsolidity/ast/ASTJsonImporter.cpp]: Why did you remove this?

File: libsolidity/ast/ASTJsonImporter.cpp [hunk 28]
```diff
@@ -88,6 +88,24 @@ SourceLocation const ASTJsonImporter::createSourceLocation(Json::Value const& _n
 }
 
 //helper functions
+ASTPointer<ASTString> ASTJsonImporter::memberAsASTString(Json::Value const& _node, string _name)
```
Comment (chriseth) [inline libsolidity/ast/ASTJsonImporter.cpp]: Please use `string const& _name`

File: libsolidity/ast/ASTJsonImporter.cpp [hunk 29]
```diff
@@ -88,6 +88,24 @@ SourceLocation const ASTJsonImporter::createSourceLocation(Json::Value const& _n
...
+{
+	return make_shared<ASTString>(member(_node, _name).asString());
```
Comment (chriseth) [inline libsolidity/ast/ASTJsonImporter.cpp]: Please make this more strict, it should require the type to be string.

File: libsolidity/ast/ASTJsonImporter.cpp [hunk 30]
```diff
@@ -88,6 +88,24 @@ SourceLocation const ASTJsonImporter::createSourceLocation(Json::Value const& _n
...
+}
+
+bool ASTJsonImporter::memberAsBool(Json::Value const& _node, string _name)
+{
+	return member(_node, _name).asBool();
```
Comment (chriseth) [inline libsolidity/ast/ASTJsonImporter.cpp]: Please make this more strict, it should require the member to be a bool.

File: libsolidity/ast/ASTJsonImporter.cpp (hunk 27 above)
Comment (djudjuu) [reply libsolidity/ast/ASTJsonImporter.cpp]: I figured it makes more sense to check for them closer to where they are actually used. 
"src" gets checked in createSourceLocation, which is called right below.

I decided to use the "id" field in the error-message that is thrown when a json-AST-Node-object does not have a field that is required (accessed via member()). That way one can directly see why the AST can not be imported. But as the member-function is called before createASTNode(), I needed to check for it earlier, so now that happens in convertJsonToASTNode(),...........aaaaand I just realized that this is not the only route createASTNode() can be called so I do need to check for it here too.

**Comment by djudjuu** [comment]
should I also move all createASTNode()- and memberX()-calls directly into the return-statement instead of saving them in a variable first? Or is this unsafe somehow?
@chriseth

File: Changelog.md [hunk 31]
```diff
@@ -1,8 +1,9 @@
-### 0.4.12 (unreleased)
+
+	### 0.4.12 (unreleased)
```
Comment (axic) [inline Changelog.md]: Typo

File: libsolidity/interface/CompilerStack.cpp [hunk 32]
```diff
@@ -782,13 +813,13 @@ string CompilerStack::createOnChainMetadata(Contract const& _contract) const
...
 	meta["version"] = 1;
-	meta["language"] = "Solidity";
+	meta["language"] = m_importedSources ? "SolidityAST" : "Solidity";
 	meta["compiler"]["version"] = VersionStringStrict;
 
 	meta["sources"] = Json::objectValue;
 	for (auto const& s: m_sources)
 	{
-		solAssert(s.second.scanner, "Scanner not available");
+		solAssert(s.second.scanner,"");
```
Comment (axic) [inline libsolidity/interface/CompilerStack.cpp]: Can you please keep this?

File: solc/CommandLineInterface.cpp [hunk 33]
```diff
@@ -770,14 +849,14 @@ bool CommandLineInterface::processInput()
...
 				successful = false;
-
-		for (auto const& error: m_compiler->errors())
-			SourceReferenceFormatter::printExceptionInformation(
-				cerr,
-				*error,
-				(error->type() == Error::Type::Warning) ? "Warning" : "Error",
-				scannerFromSourceName
-			);
+		if (!m_compiler->importedSources())
```
Comment (axic) [inline solc/CommandLineInterface.cpp]: Why is this needed?

File: libsolidity/interface/CompilerStack.cpp (hunk 32 above)
Comment (djudjuu) [reply libsolidity/interface/CompilerStack.cpp]: oops, not on purpose

File: solc/CommandLineInterface.cpp (hunk 33 above)
Comment (djudjuu) [reply solc/CommandLineInterface.cpp]: as I understood it the warnings are printed with a reference to the src-code, provided by scannerFromSourceName, which does not make sense if the src-file is a JSON

Comment (axic) [reply solc/CommandLineInterface.cpp]: I think warnings should be kept. Does it crashes or just the source references are wrong?

Comment (djudjuu) [reply solc/CommandLineInterface.cpp]: it prints out the entire AST for every warning

Comment (djudjuu) [reply solc/CommandLineInterface.cpp]: I could add a flag to the printExceptionInformation that prevents the printout of the src...

Comment (djudjuu) [reply solc/CommandLineInterface.cpp]: hmm, I tried that and it means changing a whole bunch of function Calls, for lots of them I did not know exactly whether they could also be called in the context of an imported AST...So I decided to just leave it in.

**Comment by chriseth** [comment]
@djudjuu you can move the creation calls directly into the return statement.

File: libsolidity/interface/CompilerStack.cpp [hunk 34]
```diff
@@ -632,8 +663,8 @@ void CompilerStack::resolveImports()
 			if (ImportDirective const* import = dynamic_cast<ImportDirective*>(node.get()))
 			{
 				string const& path = import->annotation().absolutePath;
-				solAssert(!path.empty(), "");
-				solAssert(m_sources.count(path), "");
+				solAssert(!path.empty(), "sdfsdfsdf");
+				solAssert(m_sources.count(path), "sdfsf");
```
Comment (axic) [inline libsolidity/interface/CompilerStack.cpp]: If you need this can you insert meaningful messages? :)

File: libsolidity/interface/CompilerStack.cpp [hunk 35]
```diff
@@ -800,12 +831,12 @@ string CompilerStack::createOnChainMetadata(Contract const& _contract) const
 				"bzzr://" + toHex(dev::swarmHash(s.second.scanner->source()).asBytes())
 			);
 		}
+
 	}
 	meta["settings"]["optimizer"]["enabled"] = m_optimize;
 	meta["settings"]["optimizer"]["runs"] = m_optimizeRuns;
 	meta["settings"]["compilationTarget"][_contract.contract->sourceUnitName()] =
 		_contract.contract->annotation().canonicalName;
-
```
Comment (axic) [inline libsolidity/interface/CompilerStack.cpp]: Extra whitespace changes.

File: libsolidity/interface/CompilerStack.cpp [hunk 36]
```diff
@@ -800,12 +831,12 @@ string CompilerStack::createOnChainMetadata(Contract const& _contract) const
 				"bzzr://" + toHex(dev::swarmHash(s.second.scanner->source()).asBytes())
 			);
 		}
+
```
Comment (axic) [inline libsolidity/interface/CompilerStack.cpp]: Extra whitespace changes.

File: libsolidity/interface/Utils.h [hunk 37]
```diff
@@ -38,6 +38,9 @@ struct UnimplementedFeatureError;
 #define solAssert(CONDITION, DESCRIPTION) \
 	assertThrow(CONDITION, ::dev::solidity::InternalCompilerError, DESCRIPTION)
 
+#define astAssert(CONDITION, DESCRIPTION) \
+	assertThrow(CONDITION, ::dev::solidity::InternalCompilerError, DESCRIPTION)
```
Comment (axic) [inline libsolidity/interface/Utils.h]: This is not using the `InvalidAstError`.

Comment (axic) [reply libsolidity/interface/Utils.h]: Thought it would be nicer to have proper error reporting, such as an *AST analyzer/validator* class which uses the `ErrorReporter`.

File: solc/CommandLineInterface.cpp [hunk 38]
```diff
@@ -760,8 +769,78 @@ bool CommandLineInterface::processInput()
...
 			m_compiler->setRemappings(m_args[g_argInputFile].as<vector<string>>());
-		for (auto const& sourceCode: m_sourceCodes)
-			m_compiler->addSource(sourceCode.first, sourceCode.second);
+
+		if (m_args.count(g_argImportAst))
+		{
+			//read file contents, which either are
+			// a) a json-file with multiple sources (as produced by --combined-json)
+			// b) a literal metadata-json-file with SolidityAST as language (as produced by --metadata --metadata-literal --import-ast)
+			// c) a json-file with one source only (as produced --ast-combined-json)
```
Comment (axic) [inline solc/CommandLineInterface.cpp]: This seems quite compliacted. If we really want to keep this kind of behaviour it may make sense moving it into a helper function.

Comment (djudjuu) [reply solc/CommandLineInterface.cpp]: I thought that it does not make sense to restrict the import to only work for one specific kind of JSON-file, moving it in a helper function does clean it up a lot though :)

**Comment by axic** [comment]
Closing in favour of #2500 (but don't delete the branch, we can reopen this if #2500 stalls).
